pkg/controller/unifiedpushserver/unifiedpushserver_controller.go

```diff
@@ -100,53 +138,185 @@ func (r *ReconcileUnifiedPushServer) Reconcile(request reconcile.Request) (recon
 		return reconcile.Result{}, err
 	}
 
-	// Define a new Pod object
-	pod := newPodForCR(instance)
+	persistentVolumeClaim, err := newPostgresqlPersistentVolumeClaim(instance)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
 
 	// Set UnifiedPushServer instance as the owner and controller
-	if err := controllerutil.SetControllerReference(instance, pod, r.scheme); err != nil {
+	if err := controllerutil.SetControllerReference(instance, persistentVolumeClaim, r.scheme); err != nil {
 		return reconcile.Result{}, err
 	}
 
-	// Check if this Pod already exists
-	found := &corev1.Pod{}
-	err = r.client.Get(context.TODO(), types.NamespacedName{Name: pod.Name, Namespace: pod.Namespace}, found)
+	// Check if this PersistentVolumeClaim already exists
+	foundPersistentVolumeClaim := &corev1.PersistentVolumeClaim{}
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: persistentVolumeClaim.Name, Namespace: persistentVolumeClaim.Namespace}, foundPersistentVolumeClaim)
 	if err != nil && errors.IsNotFound(err) {
-		reqLogger.Info("Creating a new Pod", "Pod.Namespace", pod.Namespace, "Pod.Name", pod.Name)
-		err = r.client.Create(context.TODO(), pod)
+		reqLogger.Info("Creating a new PersistentVolumeClaim", "PersistentVolumeClaim.Namespace", persistentVolumeClaim.Namespace, "PersistentVolumeClaim.Name", persistentVolumeClaim.Name)
+		err = r.client.Create(context.TODO(), persistentVolumeClaim)
 		if err != nil {
 			return reconcile.Result{}, err
 		}
+	} else if err != nil {
+		return reconcile.Result{}, err
+	}
 
-		// Pod created successfully - don't requeue
-		return reconcile.Result{}, nil
+	postgresqlDeployment, err := newPostgresqlDeployment(instance)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Set UnifiedPushServer instance as the owner and controller
+	if err := controllerutil.SetControllerReference(instance, postgresqlDeployment, r.scheme); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Check if this Deployment already exists
+	foundPostgresqlDeployment := &appsv1.Deployment{}
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: postgresqlDeployment.Name, Namespace: postgresqlDeployment.Namespace}, foundPostgresqlDeployment)
+	if err != nil && errors.IsNotFound(err) {
+		reqLogger.Info("Creating a new Deployment", "Deployment.Namespace", postgresqlDeployment.Namespace, "Deployment.Name", postgresqlDeployment.Name)
+		err = r.client.Create(context.TODO(), postgresqlDeployment)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
 	} else if err != nil {
 		return reconcile.Result{}, err
 	}
 
-	// Pod already exists - don't requeue
-	reqLogger.Info("Skip reconcile: Pod already exists", "Pod.Namespace", found.Namespace, "Pod.Name", found.Name)
-	return reconcile.Result{}, nil
-}
+	postgresqlService, err := newPostgresqlService(instance)
```

Comment:
IHMO the DB should not be called/used in this way and I'd like to recommend do a CR for the database which should have its own controller as was done [here](https://github.com/aerogear/mobile-security-service-operator/tree/master/pkg/controller). 

## Why? 

According to designed proposed for operator-framework, each thing should be a CR/type of resource definition. In this way, UPS and Database in my point of view cannot be the same type of object/resource.

The App/Service is one CR as the Database is another. Note that each CR means one type of resource. If you follow the internal course materials and/or its docs as the examples found in the operatorHub we will see this design properly in action. 

I'd like to highlight that this approach in my point of view goes against the concept to do a design driven by the domain as well and that how much more atomic are the objects/types and implementations better are to keep them maintained.

As you can check [here](https://github.com/operator-framework/operator-sdk/issues/1383) do many CRDs and use one controller according to the teammates of the framework also goes against to design goals for the controller-runtime.

## Examples

* https://github.com/coreos/prometheus-operator
The Prometheus has one type/ CR for each thing. 

* https://operatorhub.io/operator/alpha/aqua-operator.v0.0.1
* https://github.com/aquasecurity/aqua-operator/tree/master/deploy/crds
This one has the service and database: ( AquaDatabase CR, AquaServer CR, AquaGateway CR and etc )

* https://operatorhub.io/operator/alpha/aws-service-operator.v0.0.1
This one is from AWS in HELM and has one CR for each thing as well ( DynamoDB, ECRRepository and etc .. )

* https://github.com/awslabs/aws-service-operator/tree/master/examples
Etcd is an excellent project to follow like example and is doing one CR for each thing ( etcd Cluster, etcd Backup, etcd Restore )

* https://github.com/coreos/etcd-operator/tree/master/example
* https://operatorhub.io/operator/clusterwide-alpha/etcdoperator.v0.9.4-clusterwide

Please, feel free to check the operators available in https://operatorhub.io/ 

## Specific reasons to keep each impl in a CRs regards our business

It is clear that we have the intention to remove the DB since providing the DB as service is not our goal. May in the future we would like to change it to use a community and/or a Redhat operator for the database then following some extra argumentations.

As it was implemented/designed we are not allowed to:

- Just not install the type DB and use a community operator if we wish
- Remove the DB from the project easily since is not very clear what should be removed or not because all code is impl together.

Reply:
You might be right about this, but I'm not convinced (at least not yet).

In my opinion, the PostgreSQL server is just an implementation detail. We don't want people to have to configure a PostgreSQL through a CR so that then they can configure UPS to use that in their UPS CR: all they want is UPS, and we don't want them to touch this PostgreSQL themselves.

Yes, in future it would be great if all of the DB management stuff was handled by a PostgreSQL operator, but I'm not convinced that this means we need to have a separate controller & CRD for that in this operator. At that point, I don't think it will be any more work really to remove it from here as it would to remove a separate controller & CRD (also, not sure what would happen to our custom PostgreSQL CRs that people might have created in their projects before then).

@wei-lee @maleck13 do either of you have opinions or thoughts on what's best here, if any?

Reply:
Hi @grdryn,

IMHO:

- In order to have a CR with DB, you can basically cut and past the MSS-Operator solution since it is the same need which means that shows be **really small effort**.  
- Do not put all impl in the same CR doesn't make the things harder just the opposite. 
- In MSS Oper we use make commands to make install with just one command and if/when
we provide it by OLM or RHMI, for example, we can just declare the CRs dependency according to the guidelines of the framework adopted. 
- Not follow the framework guidelines can lead us to start to face unexpected scenarios and start to impl workarounds which will make harder put the things in the trail in the future
- I do not agree with the approach of starting new projects from scratch doing what was not supposed to be done and I think it will lead us a similar situation faced already where will be hard to improve the projects or keep them maintained because they grew unduly which means highly effort in the future. 

It is my recommendation/suggestion and point of view, however, it is up to you folks.

Reply:
I have the same view: we should allow users to configure certain aspects of the db through the same CR (e.g. size of the volume), but I don't really think we need a separate CR and have a separate controller to manage the database, and it is not really the job of the operator for this service.  If we are going to have another CR just for the db, then why not just use a separate db operator instead?

Reply:
Hi @wei-lee,
 
If the DB is not part of the service how we can add it in the middle of the impl of the service and defined it as part of the service provided by this CR? 

Following some comments inline.

>  we should allow users to configure certain aspects of the db through the same CR (e.g. size of the volume),

We do that in the MSS Operator. The Service CR creates a ConfigMap with all values that the DB should use and what and where we will allow the customer to define the specs has no relation with the fact of we following the guideline and have a CR for each type of thing as the other operators.

See: 

> If we are going to have another CR just for the db, then why not just use a separate db operator instead?  

Definitely, I believe that we should do that and this is one of the many reasons in POV for we NOT impl the DB as part of the Service. Keep the database as part of the service will not allow us to remove or change it with a small impact and effort and will make really harder we are able to rip off the DB in the future. :-(  

So, why not use now an operator DB? 

* Unfurtunully, as we spoke before has not a PostgreSQL operator available which will attend our needs so far. However, is highly like that it can be provided by Red Hat soon since for OCP 3.11 we have the PostgreSQL card/image in the catalogue.
* So, to follow this approach we would need impl our PostgreSQL oper which for now at least show higher effort to get done than just cut and paste the code impl. Provide and be responsible to maintain a DB operator shows not be our goal too.
* If we check that for our operators we will start to impl to many features regards to DB so we will also be able to easily remove the CR/CRD from the project and use it to create our DB operator which can be used for all services. 

I hope that I could clarify my POV and that information can be helpful.

Reply:
My 2 cents here. 

- ideally Postgres would be managed by another operator and UPS would depend on that operator and create a CR for the postgres it needed. Any additional info / config would be done via that CR
- pragmatically, this is not the case and wont be for some time. So I would be in favour of not exposing Database info in the CR (it is important to think of these as an API, the more you expose the more you have to support and be back compat with). Having a default setup that suits 95% of usecases is a good place to start and then wait for the additional use cases edge cases to filter back from actual use cases.

Reply:
Thanks for the feedback everyone. I'm going to leave it as-is for the moment and merge this today.

For now until there's a postgresql operator that we can call out to, I don't think it's even a good idea to expose the db volume size through the CR, since that might give the impression that it could be tweaked (e.g. someone tries to _increase_ it from 10Gi to 20Gi after a few months, and expects something to happen...).

Reply:
Hi all, I believe we expose these in the MSS Operator. Should we also change it there? cc @davidffrench @camilamacedo86

Reply:
@laurafitzgerald You are correct and we also have an individual CR specifically for the DB. This won't be changed for us.

I agree with Craig, what I would like to see is for these values to be optional in our db CR spec with default values defined in the operator when these are not provided.